core/frontend/src/main/scala/io/udash/properties/Validator.scala

```diff
@@ -2,20 +2,41 @@ package io.udash.properties
 
 import scala.concurrent.{ExecutionContext, Future}
 
+trait ValidationError {
+  def message: String
+}
+case class DefaultValidationError(override val message: String) extends ValidationError
+
 sealed trait ValidationResult
 case object Valid extends ValidationResult
-case class Invalid(errors: Seq[String]) extends ValidationResult
+case class Invalid[ErrorType <: ValidationError](errors: Seq[ErrorType]) extends ValidationResult
+object Invalid {
+  def apply[ErrorType <: ValidationError](error: ErrorType, errors: ErrorType*): Invalid[ErrorType] =
+    Invalid(error +: errors.toSeq)
+
+  def apply(error: String, errors: String*): Invalid[ValidationError] =
+    this((error +: errors).map(DefaultValidationError))
+}
 
-trait Validator[T] {
-  def apply(element: T)(implicit ec: ExecutionContext): Future[ValidationResult]
+trait Validator[ArgumentType] {
+  def apply(element: ArgumentType)(implicit ec: ExecutionContext): Future[ValidationResult]
 }
 
 object Validator {
+  class FunctionValidator[ArgumentType](f: (ArgumentType) => ValidationResult) extends Validator[ArgumentType] {
+    override def apply(element: ArgumentType)(implicit ec: ExecutionContext): Future[ValidationResult] =
+      Future(f(element))(ec)
+  }
+
+  def apply[ArgumentType](f: (ArgumentType) => ValidationResult): Validator[ArgumentType] =
+    new FunctionValidator(f)
+
   implicit class FutureOps[T](private val future: Future[T]) extends AnyVal {
-    def foldValidationResult(implicit ev: T =:= Seq[ValidationResult], ec: ExecutionContext): Future[ValidationResult] = {
+    def foldValidationResult[ErrorType <: ValidationError](implicit ev: T =:= Seq[ValidationResult],
```

Comment:
ErrorType is not used anywhere
I've also commited a single traversal version of this method, check if you like it better ;)